### PR TITLE
Add optional parameter called Source for module win_feature

### PR DESCRIPTION
**Pull request by DigitalDrifta**
Hi,

As per bug/feature request I posted yesterday: https://github.com/ansible/ansible-modules-core/issues/1917

I have made the changes. Please let me know if it's acceptable.

Thanks,
Dean

### CODE REVIEW

**Comment by bcoca** [comment]
cc @cchurch @jhawkesworth @trondhindenes @angstwad

**Comment by trondhindenes** [comment]
Good idea! I won't be able to test this until thursday, but I'll get on it as soon as I can.

**Comment by jhawkesworth** [comment]
I agree this is a good idea but I have a couple of thoughts...

Is it worth putting something in the docs about why the source parameter is needed.  Its one of those things that you aren't likely to know about until you hit the problem.

Just wondered whether it would be worth tweaking the logic slightly so it tries the running image first (i.e. without specifying a -source) and then only if that fails will it try the source specified in the module parameters.  Or perhaps have the ability to try using the image first as another parameter.

The reason I suggest this is it allows the module to be called with the same parameters in all environments (in development you may have MSDN vms which have the features included in the image, whereas in production you may be using, say volume licensed editions of windows, which may not have the features in the image).   I use a custom version of win_feature at the moment for the reason described above.

**Comment by roligov** [comment]
From my limited testing, if you pass a source parameter , windows will attempt to install it from local first. From the looks of it I think if it cannot find the packages locally, it checks to see if there is a source, checks to see if it's valid then looks for the packages on that source. I confirmed this in the documentation for Install-WindowsFeature:

https://technet.microsoft.com/en-us/library/jj205467(v=wps.630).aspx
The path specified in this parameter is only used if the command cannot find feature files in the local side-by-side store of the specified target computer or VHD. The command searches for feature files in the following order:
 1) On the target computer or offline VHD.
 2) Path specified as the value of this parameter. If you add a UNC path, verify that the computer account of the target server has Read permissions on the share. The computer account should be in one of the following formats: DOMAIN\SERVERNAME$ or SERVER$
 3) Repository path specified by the Group Policy Object (GPO), Specify settings for optional component installation and component repair, located in Computer Configuration/Administrative Templates/System in Local Group Policy Editor. This Group Policy setting controls the following Windows Registry setting: HKEY_LOCAL_MACHINE\Software\Microsoft\Windows\CurrentVersion\Policies\Servicing\LocalSourcePath.
 4) Windows Update.
This parameter is optional.

Regards,
Dean

**Comment by jhawkesworth** [comment]
Thanks, @DigitalDrifta it sounds like I should be able to drop my custom module with this.
That's slightly surprising behaviour to me, so ideally I'd like it mentioned in the module docs that source only used if not present in the running image.
I looked over the code and it appears fine, just want to test before I upvote this.
Its on my list, but not sure when I can get to it at the moment.

Jon

**Comment by DigitalDrifta** [comment]
Sorry, looks like I've signed up with another email address in the past. roligov is my account too :)

Thanks @jhawkesworth! No rush, I've updated it on my Ansible Server so I'm good for now.

**Comment by trondhindenes** [comment]
@DigitalDrifta I take it you're not a huge fan of putting curly braces on newline since you've changed every single if statement in the code along with your additions :-)

**Comment by roligov** [comment]
Sorry about that @trondhindenes. Powershell Studio 2015 automatically corrects the formatting for me when I paste any code in there, and that is the format I use at work as well, I find it easier to read. You can change it if need be :)

**Comment by trondhindenes** [comment]
I don't mind. I tend do use newline since that's what c#/VS uses, but we're different. No beef, I just thought it was funny :-)

File: windows/win_feature.py [hunk 1]
```diff
@@ -68,6 +68,15 @@
...
     default: null
     aliases: []
+  source:
+    description:
+      - Specify a source to install the feature from
+    required: false
+    choices: 
+      - {driveletter}:\sources\sxs
+      - \\{IP}\Share\sources\sxs
+    default: null
```
Comment (petemounce) [inline windows/win_feature.py:78]: When `required: false`, `default: null` is unnecessary (https://docs.ansible.com/ansible/developing_modules.html#module-checklist)

File: windows/win_feature.py [hunk 2]
```diff
@@ -68,6 +68,15 @@
...
+    aliases: []
```
Comment (petemounce) [inline windows/win_feature.py:79]: Unnecessary (https://docs.ansible.com/ansible/developing_modules.html#module-checklist)

**Comment by petemounce** [comment]
I don't know enough about the semantics to do more than the "code looks fine" type of review, sorry.

File: windows/win_feature.ps1 [hunk 3]
```diff
@@ -47,40 +47,52 @@ Elseif (!$params.state) {
...
 }
-Else
-{
+Else {
     $includemanagementtools = $false
 }
 
+If ($params.source) {
+    $source = $params.source.ToString().ToLower()
+    If (!Test-Path $source) {
```
Comment (trondhindenes) [inline windows/win_feature.ps1:70]: I'm getting an execution error from "!test-path". You probably need to specify it as if (!(test-path $source)) {

**Comment by gregdek** [comment]
Thanks @DigitalDrifta for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by robynbergeron** [comment]
@DigitalDrifta A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by jhawkesworth** [comment]
It would be a shame to miss this, its a good idea.  

Pinging @roligov just in case you are not monitoring @DigitalDrifta email at the moment (as per above both being the same person).

**Comment by trondhindenes** [comment]
@jhawkesworth if we dont hear anything from submitter it should be easy to just add the proposed changes in a new pr. I can do that in a few days if nothing happens here.

**Comment by DigitalDrifta** [comment]
@jhawkesworth Apologies for the delayed response. Unfortunately I've had some personal things come up in the last few weeks and haven't had much time to focus on much else. I'll try and resolve the issues by next week Friday. If someone else completes I will have no issue with that either. Sorry again.

Regards.

**Comment by trondhindenes** [comment]
I created a separate PR containing the same functionality, and cleaned up the code a bit at the same time:
PR #2783

My suggestion is that we close this one and consider merging #2783 instead.

**Comment by jhawkesworth** [comment]
No worries @DigitalDrifta its understood that we do things when we can.
Thanks @trondhindenes for creating the new PR.  I have tested it successfully today so I agree that this one should close and persue this via https://github.com/ansible/ansible-modules-core/pull/2783

All the best,

Jon

**Comment by nitzmahone** [comment]
superseded by #2783, closing

**Pull request closed without merging**